File: frontend/projects/storage/src/lib/entities/storage-node.ts
export type StorageNodeType = 'FILE' | 'DIRECTORY' | 'NONE';

export interface StorageNode {
  path: string;
  type: StorageNodeType;
  depth: number;
  length: number;
  lastModified: number;
}

export const ROOT_NODE: StorageNode = {
  path: '',
  type: 'DIRECTORY',
  depth: -1,
  length: 0,
  lastModified: 0,
};

File: backend/commons/storage/storage-entity/src/main/java/com/kraken/storage/entity/StorageNode.java
package com.kraken.storage.entity;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Builder;
import lombok.Value;

import static java.util.Objects.requireNonNull;

@Value
@Builder(toBuilder = true)
public class StorageNode {

  String path;
  StorageNodeType type;
  Integer depth;
  Long length;
  Long lastModified;

  @JsonCreator
  StorageNode(
      @JsonProperty("path") final String path,
      @JsonProperty("type") final StorageNodeType type,
      @JsonProperty("depth") final Integer depth,
      @JsonProperty("length") final Long length,
      @JsonProperty("lastModified") final Long lastModified) {
    super();
    this.path = requireNonNull(path);
    this.type = requireNonNull(type);
    this.depth = requireNonNull(depth);
    this.length = requireNonNull(length);
    this.lastModified = requireNonNull(lastModified);
  }


  @JsonIgnore
  public boolean notRoot() {
    return !this.path.isEmpty();
  }
}
